src/xdist/dsession.py

```diff
@@ -182,6 +182,15 @@ def worker_workerfinished(self, node):
             self.shouldstop = f"{node} received keyboard-interrupt"
             self.worker_errordown(node, "keyboard-interrupt")
             return
+
+        shouldfail = node.workeroutput["shouldfail"]
+        shouldstop = node.workeroutput["shouldstop"]
+        for shouldx in [shouldfail, shouldstop]:
+            if shouldx:
+                self.shouldstop = shouldx
+                self._active_nodes.remove(node)
```

Comment:
Let's write the code such that this line is not repeated below. Can use `for-break-else` or whatever way you like.